src/gather.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+import {Tensor, sizeOfShape} from './lib/tensor.js';
+import {validateGatherParams} from './lib/validate-input.js';
+
+/**
+ * Gather values of the input tensor along an axis according to the indices.
+ * @param {Tensor} input
+ * @param {Tensor} indices
+ * @param {MLGatherOptions} [options]
+ * @return {Tensor}
+ */
+export function gather(input, indices, {axis = 0} = {}) {
+  validateGatherParams(...arguments);
+  const shapeInput = input.shape;
+
+  // set outputShape following Spec Algorithm
+  // let dimCount = 0
+  // let rankOutput = 0;
+  // let shapeOutput = [];
+  // for (dimCount = 0; dimCount < shapeInput.length; dimCount++) {
+  //   if (dimCount === axis) {
+  //     break;
+  //   } else {
+  //     shapeOutput[dimCount] = shapeInput[dimCount];
+  //   }
+  // }
+  // rankOutput = dimCount;
+  // for (dimCount = 0; dimCount < indices.shape.length; dimCount++) {
+  //   shapeOutput[rankOutput + dimCount] = indices.shape[dimCount];
+  // }
+  // rankOutput = rankOutput + dimCount;
+  // for (dimCount = 0; dimCount < shapeInput.length; dimCount++) {
+  //   if (dimCount <= axis) {
+  //     continue;
+  //   } else {
+  // here index shoud be set to (rankOutput + dimCount - axis - 1)
+  //     shapeOutput[rankOutput + dimCount - axis - 1] = shapeInput[dimCount];
+  //   }
+  // }
+
+  // optimized set outputShape using JavaScipt slice and concat
+  const shapeOutput = shapeInput.slice(0, axis).concat(indices.shape, shapeInput.slice(axis + 1));
+  const output = new Tensor(shapeOutput);
+
+  for (let i = 0; i < sizeOfShape(shapeOutput); ++i) {
+    // output[i, j, k, ...] = input[index[i, j, k, ...], j, k, ...] // if axis == 0
+    // output[i, j, k, ...] = input[i, index[i, j, k, ...], k, ...] // if axis == 1
+    // output[i, j, k, ...] = input[i, j, index[i, j, k, ...], ...] // if axis == 2
+    const loc = output.locationFromIndex(i);
+    const indicesLoc = loc.slice(axis, axis + indices.rank);
+    const selectedInputLoc = loc.slice(0, axis)
+        .concat(indices.getValueByLocation(indicesLoc), loc.slice(axis + indices.rank));
```

Comment:
Just a FYI, the current implementation in DML will not throw if the values in indices are out of bounds. 

The open spec issue for out-of-bounds handling: https://github.com/webmachinelearning/webnn/issues/486

Reply:
Thanks for the pointer. webnn-baseline implementation aims at "doing" right thing ((non out-of-bounds access)) and generating the expected test data. 

There is a separate task that explicitly checks the out-of-bounds access tracking by issue: https://github.com/webmachinelearning/webnn/issues/244#issuecomment-1820092158.